### PR TITLE
Explore: Improve local storage error handling when rich history is added

### LINKED ISSUE #38814: Explore: Query cannot be run when rich history exceeded local storage quota
**What happened**:

It seems to be rare but may happen in some cases. Explore crashes if richHistory has reached the limit of local storage and the query is run:

https://user-images.githubusercontent.com/745532/131815549-a76ecbb8-d44d-4a59-8f7a-0b4b27d13e82.mov

**What you expected to happen**:

An alert message could be displayed instead indicating that the rich history limit was hit with a possibility to clear the history.

**How to reproduce it (as minimally and precisely as possible)**:

Add this snippet before the history is saved in try/catch block here: https://github.com/grafana/grafana/blob/v8.1.2/public/app/core/utils/richHistory.ts#L77

```typescript
    for (let i = 0; i < 1000; i++) {
      updatedHistory = updatedHistory.concat({
        queries: newQueriesToSave,
        ts,
        datasourceId,
        datasourceName,
        starred,
        comment,
        sessionName,
      });
    }
```

Then run the query a few times.

**Anything else we need to know?**:

**Environment**:
- Grafana version: 8.1.2
- Data source type & version: Any
- OS Grafana is installed on: Mac
- User OS & Browser: macOS Big Sur
- Grafana plugins: None
- Others: N/A

### PR DESCRIPTION
**Pull request by ifrost**
**What this PR does / why we need it**:

We have recently fixed displaying error message when local storage quota is exceeded after adding a new rich history item. (https://github.com/grafana/grafana/pull/39766)

This however has at least two problems:
* The error is shown every time a query is run
* The storage is not cleaned

I though about cleaning rich history when quota is exceeded but the problem is that the quota is shared for all items in the storage. It might be the case that the storage is full because of other local storage items and it's not rich history that is causing problems per se. I'm not sure if there's a clean way to clean the storage. Note that we already have retention period that cleans up old items. Also this problem should be fixed once Team History is implemented.

To improve UX I thought we could:
* Display the error message only once per session - this should make the error less annoying
* On top of the retention period add maximum number of entries we keep in the storage - this should make the error happen less often.

Let me know what you think. If you agree with the approach I can add some tests.

**Which issue(s) this PR fixes**:

Fixes #38814

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/core/utils/richHistory.ts [hunk 1]
```diff
@@ -52,7 +55,8 @@ export function addToRichHistory(
   /* Keep only queries, that are within the selected retention period or that are starred.
    * If no queries, initialize with empty array
    */
-  const queriesToKeep = richHistory.filter((q) => q.ts > retentionPeriodLastTs || q.starred === true) || [];
+  const queriesToKeep =
+    richHistory.filter((q) => q.ts > retentionPeriodLastTs || q.starred === true).splice(0, MAX_HISTORY_ITEMS) || [];
```
Comment (ivanahuckova) [inline public/app/core/utils/richHistory.ts:59]: This way, we always delete the freshly run query - we are not able to add any new query. Shouldn't we delete the last unstarred query in the array? I mean this works as well, just asking why this solution was preferred.

**Review by ivanahuckova — COMMENTED** [review]
Left 1 comment below. Otherwise looks and works great, just wanted to discuss delete logic.

File: public/app/core/utils/richHistory.ts (hunk 1 above)
Comment (ifrost) [reply public/app/core/utils/richHistory.ts:59]: Good point, that should be better!

- [ ] rotate unstarred history items when reaching the limit

**Review by ivanahuckova — APPROVED** [review]
Yes, nice! 🤩 I left 1 last comments/question, that is definitely non-blocking. Just throwing it out there. 😂

**Pull request merged**